// @uiw/react-markdown-preview/lib/esm/index.js
import _extends from "@babel/runtime/helpers/extends";
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _objectSpread from "@babel/runtime/helpers/objectSpread2";
import React, { useImperativeHandle } from 'react';
import ReactMarkdown from 'react-markdown';
import gfm from 'remark-gfm'; // @ts-ignore

import slug from 'remark-slug'; // @ts-ignore

import headings from 'remark-autolink-headings';
import rehypeRaw from 'rehype-raw';
import rehypePrism from '@mapbox/rehype-prism';
import rehypeRewrite from 'rehype-rewrite';
import "./styles/markdown.css";
import "./styles/markdowncolor.css";

var rehypeRewriteHandle = function rehypeRewriteHandle(node, index, parent) {
  if (node.type === 'element' && parent.type === 'root' && /h(1|2|3|4|5|6)/.test(node.tagName) && index !== 0) {
    var child = node.children && node.children[0] ? node.children[0] : null;

    if (child && child.properties && child.properties.ariaHidden === 'true') {
      child.properties = _objectSpread({
        class: 'anchor'
      }, child.properties);
      child.children = [{
        type: 'element',
        tagName: 'svg',
        properties: {
          class: 'octicon octicon-link',
          viewBox: '0 0 16 16',
          version: '1.1',
          width: '16',
          height: '16',
          ariaHidden: 'true'
        },
        children: [{
          type: 'element',
          tagName: 'path',
          properties: {
            fillRule: 'evenodd',
            d: 'M7.775 3.275a.75.75 0 001.06 1.06l1.25-1.25a2 2 0 112.83 2.83l-2.5 2.5a2 2 0 01-2.83 0 .75.75 0 00-1.06 1.06 3.5 3.5 0 004.95 0l2.5-2.5a3.5 3.5 0 00-4.95-4.95l-1.25 1.25zm-4.69 9.64a2 2 0 010-2.83l2.5-2.5a2 2 0 012.83 0 .75.75 0 001.06-1.06 3.5 3.5 0 00-4.95 0l-2.5 2.5a3.5 3.5 0 004.95 4.95l1.25-1.25a.75.75 0 00-1.06-1.06l-1.25 1.25a2 2 0 01-2.83 0z'
          }
        }]
      }];
    }
  }
};

export default /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _ref = props || {},
      className = _ref.className,
      source = _ref.source,
      style = _ref.style,
      onScroll = _ref.onScroll,
      onMouseOver = _ref.onMouseOver,
      _ref$warpperElement = _ref.warpperElement,
      warpperElement = _ref$warpperElement === void 0 ? {} : _ref$warpperElement,
      other = _objectWithoutProperties(_ref, ["className", "source", "style", "onScroll", "onMouseOver", "warpperElement"]);

  var mdp = /*#__PURE__*/React.createRef();
  useImperativeHandle(ref, function () {
    return _objectSpread(_objectSpread({}, props), {}, {
      mdp: mdp
    });
  }, [mdp, props]);
  var cls = "wmde-markdown wmde-markdown-color ".concat(className || '');
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: mdp,
    onScroll: onScroll,
    onMouseOver: onMouseOver
  }, warpperElement, {
    className: cls,
    style: style
  }), /*#__PURE__*/React.createElement(ReactMarkdown, _extends({}, other, {
    rehypePlugins: [[rehypePrism, {
      ignoreMissing: true
    }], [rehypeRewrite, rehypeRewriteHandle], rehypeRaw].concat(_toConsumableArray(other.rehypePlugins || [])),
    remarkPlugins: [slug, headings].concat(_toConsumableArray(other.remarkPlugins || []), [gfm]),
    children: source || ''
  })));
});
