// @uiw/react-md-editor/lib/cjs/components/Toolbar/Child.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Child;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _ = _interopRequireDefault(require("./"));

var _Context = require("../../Context");

function Child(props) {
  var _ref = props || {},
      prefixCls = _ref.prefixCls,
      groupName = _ref.groupName,
      commands = _ref.commands,
      children = _ref.children;

  var _useContext = (0, _react.useContext)(_Context.EditorContext),
      _useContext$barPopup = _useContext.barPopup,
      barPopup = _useContext$barPopup === void 0 ? {} : _useContext$barPopup;

  return (0, _react.useMemo)(function () {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixCls, "-toolbar-child ").concat(groupName && barPopup[groupName] ? 'active' : ''),
      onClick: function onClick(e) {
        return e.stopPropagation();
      }
    }, Array.isArray(commands) ? /*#__PURE__*/_react.default.createElement(_.default, (0, _extends2.default)({
      commands: commands
    }, props, {
      height: "",
      isChild: true
    })) : children);
  }, // eslint-disable-next-line react-hooks/exhaustive-deps
  [commands, barPopup, groupName, prefixCls]);
}

module.exports = exports.default;
