// @uiw/react-markdown-preview/lib/cjs/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var _react = _interopRequireWildcard(require("react"));

var _reactMarkdown = _interopRequireDefault(require("react-markdown"));

var _remarkGfm = _interopRequireDefault(require("remark-gfm"));

var _remarkSlug = _interopRequireDefault(require("remark-slug"));

var _remarkAutolinkHeadings = _interopRequireDefault(require("remark-autolink-headings"));

var _rehypeRaw = _interopRequireDefault(require("rehype-raw"));

var _rehypePrism = _interopRequireDefault(require("@mapbox/rehype-prism"));

var _rehypeRewrite = _interopRequireDefault(require("rehype-rewrite"));

// @ts-ignore
// @ts-ignore
var rehypeRewriteHandle = function rehypeRewriteHandle(node, index, parent) {
  if (node.type === 'element' && parent.type === 'root' && /h(1|2|3|4|5|6)/.test(node.tagName) && index !== 0) {
    var child = node.children && node.children[0] ? node.children[0] : null;

    if (child && child.properties && child.properties.ariaHidden === 'true') {
      child.properties = (0, _objectSpread2.default)({
        class: 'anchor'
      }, child.properties);
      child.children = [{
        type: 'element',
        tagName: 'svg',
        properties: {
          class: 'octicon octicon-link',
          viewBox: '0 0 16 16',
          version: '1.1',
          width: '16',
          height: '16',
          ariaHidden: 'true'
        },
        children: [{
          type: 'element',
          tagName: 'path',
          properties: {
            fillRule: 'evenodd',
            d: 'M7.775 3.275a.75.75 0 001.06 1.06l1.25-1.25a2 2 0 112.83 2.83l-2.5 2.5a2 2 0 01-2.83 0 .75.75 0 00-1.06 1.06 3.5 3.5 0 004.95 0l2.5-2.5a3.5 3.5 0 00-4.95-4.95l-1.25 1.25zm-4.69 9.64a2 2 0 010-2.83l2.5-2.5a2 2 0 012.83 0 .75.75 0 001.06-1.06 3.5 3.5 0 00-4.95 0l-2.5 2.5a3.5 3.5 0 004.95 4.95l1.25-1.25a.75.75 0 00-1.06-1.06l-1.25 1.25a2 2 0 01-2.83 0z'
          }
        }]
      }];
    }
  }
};

var _default = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var _ref = props || {},
      className = _ref.className,
      source = _ref.source,
      style = _ref.style,
      onScroll = _ref.onScroll,
      onMouseOver = _ref.onMouseOver,
      _ref$warpperElement = _ref.warpperElement,
      warpperElement = _ref$warpperElement === void 0 ? {} : _ref$warpperElement,
      other = (0, _objectWithoutProperties2.default)(_ref, ["className", "source", "style", "onScroll", "onMouseOver", "warpperElement"]);

  var mdp = /*#__PURE__*/_react.default.createRef();

  (0, _react.useImperativeHandle)(ref, function () {
    return (0, _objectSpread2.default)((0, _objectSpread2.default)({}, props), {}, {
      mdp: mdp
    });
  }, [mdp, props]);
  var cls = "wmde-markdown wmde-markdown-color ".concat(className || '');
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: mdp,
    onScroll: onScroll,
    onMouseOver: onMouseOver
  }, warpperElement, {
    className: cls,
    style: style
  }), /*#__PURE__*/_react.default.createElement(_reactMarkdown.default, (0, _extends2.default)({}, other, {
    rehypePlugins: [[_rehypePrism.default, {
      ignoreMissing: true
    }], [_rehypeRewrite.default, rehypeRewriteHandle], _rehypeRaw.default].concat((0, _toConsumableArray2.default)(other.rehypePlugins || [])),
    remarkPlugins: [_remarkSlug.default, _remarkAutolinkHeadings.default].concat((0, _toConsumableArray2.default)(other.remarkPlugins || []), [_remarkGfm.default]),
    children: source || ''
  })));
});

exports.default = _default;
module.exports = exports.default;
